### PR TITLE
.Net: Enhance volatile memory connector to allow collection to be serialized/deserialized

### LINKED ISSUE #8181: .Net: Enhance volatile memory connector to allow collection to be serialized/deserialized
The scenario we would like to support is:

1. My application has a source of data that is fairly static e.g. sample data or even static for the life time of the application
2. I would like to load the data parse it and generate embeddings which are stored in volatile memory
3. The next time I run I'd like to be able to reuse the embeddings previously created

The proposal is to add functionality to the volatile memory connector so that it can serialise and deserialise it's contents to/from a stream so the data could be stored locally.

### PR DESCRIPTION
**Pull request by markwallace-microsoft**
### Motivation and Context

Closes #8181

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

**Comment by RogerBarreto** [comment]
I think we are starting to mix two things differently suggest to keep Volatile only for In-Memory processing and introduce a new `DiskMemoryStore` or `FileMemoryStore` for it.

PS: The `Volatile` name can be dropped and `Memory` literally would mean we store In-Memory (RAM).

File: dotnet/src/SemanticKernel.Core/Data/VolatileVectorStoreExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        {
+            throw new InvalidOperationException($"Collection '{collectionName}' does not exist.");
+        }
+
+        var volatileCollection = collection as VolatileVectorStoreRecordCollection<TKey, TRecord>;
+        var records = volatileCollection!.GetCollectionDictionary();
+        if (records is null)
+        {
+            throw new InvalidOperationException($"Collection '{collectionName}' is not the correct type.");
+        }
```
Comment (westey-m) [inline dotnet/src/SemanticKernel.Core/Data/VolatileVectorStoreExtensions.cs:94]: GetCollectionDictionary already throws if the collection isn't in the dictionary.  And GetCollection on line 82 throws the not correct type error if the types don't match, so this check shouldn't be required.

File: dotnet/src/SemanticKernel.Core/Data/VolatileVectorStoreExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                await Task.WhenAll(tasks).ConfigureAwait(false);
+            }
+        }
+
+        return collection;
+    }
+
+    #region private
+    /// <summary>JSON serialization configuration for readable output.</summary>
+    private static readonly JsonSerializerOptions s_jsonSerializerOptions = new() { WriteIndented = true };
```
Comment (westey-m) [inline dotnet/src/SemanticKernel.Core/Data/VolatileVectorStoreExtensions.cs:140]: This is nice for test scenarios, but if anyone wants to use this for prod scenarios, this default is not efficient.  Should we default to the most prod friendly option (not indented)?

File: dotnet/src/SemanticKernel.Core/Data/VolatileVectorStoreExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        where TKey : notnull
+        where TRecord : class
+    {
+        IVectorStoreRecordCollection<TKey, TRecord>? collection = null;
+
+        using (StreamReader streamReader = new(stream))
+        {
+            string result = streamReader.ReadToEnd();
+            var recordCollection = JsonSerializer.Deserialize<VolatileRecordCollection<TKey, TRecord>>(result);
+            if (recordCollection is not null)
```
Comment (westey-m) [inline dotnet/src/SemanticKernel.Core/Data/VolatileVectorStoreExtensions.cs:120]: Should we throw if it is null?

**Comment by markwallace-microsoft** [comment]
> I think we are starting to mix two things. Once you have a `Disk` it's not volatile anymore.
> 
> Suggest to keep Volatile only for In-Memory processing and introduce a new `DiskMemoryStore` or `FileMemoryStore` for it.



> I think we are starting to mix two things. Once you have a `Disk` it's not volatile anymore.
> 
> Suggest to keep Volatile only for In-Memory processing and introduce a new `DiskMemoryStore` or `FileMemoryStore` for it.

The only dependency is on `Stream`

File: dotnet/samples/Concepts/Memory/VolatileVectorStore_LoadData.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        // Create an embedding generation service.
+        var embeddingGenerationService = new OpenAITextEmbeddingGenerationService(
+                modelId: TestConfiguration.OpenAI.EmbeddingModelId,
+                apiKey: TestConfiguration.OpenAI.ApiKey);
+
+        // Construct a volatile vector store.
+        var vectorStore = new VolatileVectorStore();
+        var collectionName = "records";
+
+        // Path to the file where the record collection will be loaded from.
```
Comment (westey-m) [inline dotnet/samples/Concepts/Memory/VolatileVectorStore_LoadData.cs:28]:
```suggestion
        // Path to the file where the record collection will be saved to and loaded from.
```

File: dotnet/samples/Concepts/Memory/VolatileVectorStore_LoadData.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        string filePath = Path.Combine(Path.GetTempPath(), "semantic-kernel-info.json");
+        if (!File.Exists(filePath))
+        {
+            // Record collection doesn't exist so create it from the sample text file.
```
Comment (westey-m) [inline dotnet/samples/Concepts/Memory/VolatileVectorStore_LoadData.cs:32]: This comment applies to a fairly large section, but it looks like it's only for the following two lines.  It confused me for a sec, so consider making it more applicable to the local bits, e.g.
```suggestion
            // Read a list of text strings from a file, to load into a new record collection.
```

File: dotnet/src/SemanticKernel.Core/Data/VolatileVectorStoreExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    /// <summary>
+    /// Deserialize a <see cref="IVectorStoreRecordCollection{TKey, TRecord}"/> to a stream as JSON.
+    /// </summary>
+    /// <typeparam name="TKey">Type of the record key.</typeparam>
+    /// <typeparam name="TRecord">Type of the record.</typeparam>
+    /// <param name="vectorStore">Instance of <see cref="VolatileVectorStore"/> used to retrieve the collection.</param>
+    /// <param name="stream">The stream to read the serialized JSON from.</param>
+    public static async Task<IVectorStoreRecordCollection<TKey, TRecord>?> DeserializeFromJsonAsync<TKey, TRecord>(
```
Comment (westey-m) [inline dotnet/src/SemanticKernel.Core/Data/VolatileVectorStoreExtensions.cs:58]: Should we call this `DeserializerCollectionFromJsonAsync`?  E.g. If later we wanted to add a deserialize for the entire store, then I can imagine that the current name is what we would want to use.

**Review by SergeyMenshykh — APPROVED** [review]

**Pull request merged**